pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,178 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a new cache to watch on Route objects from every namespace.
+	newCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cache to the manager so that the cache is started along with the other runnables.
+	mgr.Add(newCache)
+	reconciler := &reconciler{
+		cache:     newCache,
+		Namespace: namespace,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCache),
+		handler.EnqueueRequestsFromMapFunc(reconciler.routeToIngressController)); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeToIngressController creates a reconcile.Request for all the Ingress Controllers related to the Route object.
+func (r *reconciler) routeToIngressController(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ic := range routeObject.Status.Ingress {
+		log.Info("queueing ingresscontroller", "name", ic.RouterName)
+		// Create a reconcile.Request for the Ingress Controller.
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      ic.RouterName,
+				Namespace: r.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
```

Comment:
There is a circumstance where it could be temporarily not found, e.g. when it is being deleted and recreated.  Even though this is an edge case, I don't think it is safe to delete the metrics.  We will have to continue the conversation about why the metrics should be deleted, and especially why it has to happen here.

Reply:
The reconcile loop operates based on CRUD events. If the Ingress Controller is deleted and then recreated, we'll get 2 separate events of the Ingress Controller object in addition to any other prior events. If for any prior event, during the reconcile the object is not found, the reconcile for the create event will ensure that even if the metrics is deleted, it will get added back.

Reply:
Ok - you moved the delete to the Finalizer.